app/scripts/controllers/permissions/index.js

```diff
@@ -510,6 +507,119 @@ export class PermissionsController {
     })
   }
 
+  /**
+   * Removes all known domains and their related permissions.
+   */
+  clearPermissions () {
+    this.permissions.clearDomains()
+    this.notifyAllDomains({
+      method: NOTIFICATION_NAMES.accountsChanged,
+      result: [],
+    })
+  }
+
+  /**
+   * Stores domain metadata for the given origin (domain).
+   * Deletes metadata for domains without permissions in a FIFO manner.
+   * Metadata is only deleted for domains with permissions on boot in order to
+   * prevent a degraded user experience, as metadata currently can't be
+   * requested on-demand.
+   *
+   * @param {string} origin - The origin whose domain metadata to store.
+   * @param {Object} metadata - The domain's metadata that will be stored.
+   */
+  addDomainMetadata (origin, metadata) {
+
+    const oldMetadataState = this.store.getState()[METADATA_STORE_KEY]
+    const newMetadataState = { ...oldMetadataState }
+
+    if (Object.keys(newMetadataState).length >= METADATA_STORE_MAX_SIZE) {
```

Comment:
We should only be considering the length of the metadata for sites without permissions. Otherwise things will get pretty weird for users with >100 sites connected - they'll be unable to gather more metadata, so our UI is gonna look pretty broken.

Reply:
What this function does should be clearer now that `popDomainMetadata` is inlined. 

We consider the size of the entire store, but only delete metadata for domains without permissions. Which domains have been added during this session is kept track of in the `_newMetadataOrigins` set.

We iterate over the values of `_newMetadataOrigins` and delete the metadata of the first domain without metadata that we find. We _also_ delete all origins we iterate over from `_newMetadataOrigins` because they will either be deleted entirely, or they have permissions and won't be deleted this session.

Reply:
I just don't follow why "domains that have been added during this session" is something we need to keep track of. What is a "session" in this contenxt, and why would we want to consider only these domains?

Reply:
The way I'd imagine we'd remove site metadata is to iterate over the site metadata, and remove the first one we find without permissions (as the iteration order should be guaranteed to be the same as the insertion order as of ES2015, so the first one will also be the oldest). Then we wouldn't need to track this extra state.

Reply:
Oh, but to return to the original comment in this thread:

>We consider the size of the entire store

This is the part that I think is broken; if the user has >=100 sites connected, new sites won't correctly populate the site metadata. We could fix this by only considering metadata for sites without permissions in the first place.

Reply:
> What is a "session" in this context, and why would we want to consider only these domains?

By "session", I mean one lifecycle of the MetaMask background process. So, from the time the extension background process starts until it exits.

> The way I'd imagine we'd remove site metadata is to iterate over the site metadata, and remove the first one we find without permissions

> This is the part that I think is broken; if the user has >=100 sites connected, new sites won't correctly populate the site metadata. We could fix this by only considering metadata for sites without permissions in the first place.

That's what this implementation currently does, except that it counts all domains with metadata, as opposed to only those without permissions, and it does not consider domains that ever had metadata.

I can change the trigger to only count domains without metadata, or rather the size of the `_newMetadataOrigins`.

The reason I'm not just iterating over the entire store and deleting the first thing I find is (1) I think it's inelegant (and potentially inefficient) to iterate over the entire store every time we want to delete something, and (2) we can't safely delete metadata for domains that have had permissions during a session of MetaMask, until we can request metadata on demand.

For (2), consider a user that removes permissions from a connected site, and the site's metadata gets deleted. Now the user clicks connect on that site (or opens some other tab expecting metadata), without refreshing its tab. There's now no metadata for that site, and our UI is broken.

Given all of the above, I added `_newMetadataOrigins` to keep track of domains added during the current session of MetaMask, and only delete metadata for domains without metadata in that set, in insertion order (FIFO).

I'll update the condition triggering the deletion attempt to check the size of the set instead of the size of the metadata store. That way, we should always have the metadata we need, although the UI may still break if a user leaves a tab open for a very long time and attempts to connect after opening 100 additional sites.

Reply:
This was resolved out of band.